### ISSUE TITLE
match expression falling through?

### ISSUE THREAD

**Issue by jrvidal** [rust-lang/rust#26251]
I tried this code:

```rust
fn main() {
    let x = 'a';

    match x {
            'a'...'b' if false => {
                println!("one");
            },

            'a' => {
                println!("two");
            },

            'a'...'b' => {
                println!("three");
            },

            _ => panic!("what?")
    }
}
```

I expected to get `two` but I get `three`. Is this the expected semantics?

## Meta

`rustc --version --verbose`:

rustc 1.1.0-beta (cd7d89af9 2015-05-16) (built 2015-05-16)
binary: rustc
commit-hash: cd7d89af9169885642d43597302af69f842bbd78
commit-date: 2015-05-16
build-date: 2015-05-16
host: x86_64-unknown-linux-gnu
release: 1.1.0-beta

**Comment by shepmaster** [comment]
If you increase the first range to `'a'...'c'`, then the second match arm is triggered. If you comment out either match arm with a range, then the expected match arm is triggered. Pure speculation, I wonder if there is an optimization that is reordering the match arms with the same check but different guard clauses.

### With modified range

```rust
fn main() {
    let x = 'a';

    match x {
        'a'...'c' if false => {
            println!("one");
        },

        'a' => {
            println!("two");
        },

        'a'...'b' => {
            println!("three");
        },

        _ => panic!("what?")
    }
}
```

### With commented-out following match arm

```rust
fn main() {
    let x = 'a';

    match x {
        'a'...'b' if false => {
            println!("one");
        },

        'a' => {
            println!("two");
        },

/*
        'a'...'b' => {
            println!("three");
        },
*/

        _ => panic!("what?")
    }
}
```

**Comment by Stebalien** [comment]
#18060 is similar but doesn't have guards. IMO, these should both be marked P-high, I-wrong.

**Comment by alexcrichton** [comment]
triage: I-nominated

**Comment by nikomatsakis** [comment]
cc @jakub-

**Comment by nikomatsakis** [comment]
I agree this is totally bogus... seems like the chance of someone *relying* on this behavior is... low but not non-trivial?

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by Stebalien** [comment]
jrvidal ran across this in the wild so it can't be that low.

**Comment by jrvidal** [comment]
FWIW, here is a simplified version of my use case, where I try to parse a number literal:

```rust
loop {
    match c {
        // ...
        '0'...'9' | 'a'...'f' if valid_digit(c, radix) => {/* parse digit */},
        'e' | 'f' | 's' | 'd' | 'l' if is_decimal_literal() => {/* parse exponential marker */},
        '0'...'9' | 'a'...'f' => {/* return error, bad digit */},
        _ => break
    }
}
```

**Comment by jakub-** [comment]
The root problem is the same as https://github.com/rust-lang/rust/issues/18060.

I've gotta say that I'm convinced the trans bits for `match` expressions are desperate for a rewrite that I've always hoped to take on and that I've heard @eddyb beg for as well. Sadly, time is extremely scarce for me right now so I can't promise much (oh, the bus factors…).

This, however, is extremely worrisome in that it can lead to really nasty bugs in user code.

**Comment by nrc** [comment]
triage: P-medium

also mir, apparently

**Comment by jrvidal** [comment]
I just compiled with the latest nightly and it seems to be working!

`rustc 1.12.0-nightly (0ef24eed2 2016-08-10)`
